Can now use_grab on sleepers

emmanuelbassil opened the pull request:
🆑 emmanuelbassil
tweak: Can now use grabs to put mobs in sleepers. This is already existing behavior for cryopods and bodyscanners.
tweak: Can no longer put a mob in a sleeper/cryotube/cryopod/body scanners if they are being grabbed by someone other than you.
tweak: Added eject verb to sleeper. Remember, you can use Alt+Click to eject people from sleepers.
bugfix: Fixed bug where a mob could not be buckled if mob was grabbing itself.
bugfix: Grabs and pulls now properly release when you put a mob in a sleeper, cryopod, body scanner, or cryotube
/🆑 

Adds use_grab to sleepers, cryopods, and cryotubes.
You should not still be pulling/grabbing a person once they're in there, that is now fixed.
Also prevents from moving a person into these things if they're buckled or grabbed by someone else. If too disruptive, can be removed from this PR.

SierraKomodo commented on code/game/machinery/Sleeper.dm:
```diff
@@ -192,9 +192,18 @@
 	go_in(target, user)
+	return TRUE
+
+/obj/machinery/sleeper/use_grab(obj/item/grab/grab, list/click_params) //Grab is deleted at the level of go_in if all checks are passed.
+	MouseDrop_T(grab.affecting, grab.assailant)
```
`use_grab()` must return a boolean value indicating the interaction was handled.

SierraKomodo commented on code/game/machinery/cryo.dm:
```diff
@@ -299,52 +289,70 @@
+	user.visible_message(SPAN_NOTICE("\The [user] begins placing \the [target] into \the [src]."), SPAN_NOTICE("You start placing \the [target] into \the [src]."))
+	put_mob(target, user)
 
+/obj/machinery/atmospherics/unary/cryo_cell/use_grab(obj/item/grab/grab, list/click_params) //Grab is deleted at the level of put_mob if all checks are passed.
+	MouseDrop_T(grab.affecting, grab.assailant)
```
As above

SierraKomodo commented on code/game/machinery/cryopod.dm:
```diff
@@ -437,45 +437,48 @@
-			return
-
-		attempt_enter(grab.affecting, user)
+/obj/machinery/cryopod/use_grab(obj/item/grab/grab, list/click_params) //Grab is deleted at the level of attempt_enter if all checks are passed.
+	MouseDrop_T(grab.affecting, grab.assailant)
```
As above

emmanuelbassil commented:
Fixed.
Also added remove_grabs() in mob_helpers since the same piece of code is called four times.

emmanuelbassil commented:
So forgot to do the body scanner, that's one less attack_by; and now the same rules re: grabs also apply to it.
Was also inspired by how the body scanner does sanity checks to apply them to the three others. Looks much cleaner now!

emmanuelbassil commented:
Made spacing more consistent. Will keep this in separate commit in case I broke something/I overstepped/ you want me to reverse.

emmanuelbassil commented:
Made spacing consistent.
This is ready for review again!

SierraKomodo commented on code/game/machinery/Sleeper.dm:
```diff
@@ -183,18 +183,47 @@
 	return ..()
 
+/obj/machinery/sleeper/proc/user_can_move_target_inside(mob/target, mob/user)
+	if (!istype(target))
```
- You have a `user` parameter, so use that instead of `usr`.
- Failure feedback messages should use `SPAN_WARNING()` instead of `SPAN_DANGER()`.

SierraKomodo commented on code/game/machinery/Sleeper.dm:
```diff
@@ -225,25 +254,26 @@
 
-/obj/machinery/sleeper/proc/go_in(mob/M, mob/user)
-	if(!M)
+/obj/machinery/sleeper/proc/go_in(mob/target, mob/user)
```
There's some inconsistancy in your returns for this proc - Some of them return a boolean value, some return nothing.

SierraKomodo commented on code/game/machinery/Sleeper.dm:
```diff
@@ -330,3 +354,17 @@
 		available_chemicals -= antag_chemicals
 	return 1
+
+/obj/machinery/sleeper/AltClick(mob/user)
+	if (CanDefaultInteract(user))
+		go_out()
+		return TRUE
+	return ..()
+
+/obj/machinery/sleeper/verb/eject()
+	set name = "Eject Sleeper"
+	set category = "Object"
+	set src in oview(1)
+	if (usr.stat != 0)
+		return
+	go_out()
```
The alt+click and eject verb conditions should be the same.

SierraKomodo commented on code/game/machinery/bodyscanner.dm:
```diff
@@ -126,23 +98,55 @@
 		icon_state = "body_scanner_2"
 
-//Like grap-put, but for mouse-drop.
+/obj/machinery/bodyscanner/proc/user_can_move_target_inside(mob/target, mob/user)
```
As with the sleeper, use `user` and `SPAN_WARNING()` in this proc.

SierraKomodo commented on code/game/machinery/bodyscanner.dm:
```diff
@@ -126,23 +98,55 @@
+		return FALSE
+	return TRUE
+
 /obj/machinery/bodyscanner/MouseDrop_T(mob/target, mob/user)
```
Incosnsitant return values here as well.

SierraKomodo commented on code/game/machinery/cryo.dm:
```diff
@@ -299,52 +289,80 @@
-/obj/machinery/atmospherics/unary/cryo_cell/MouseDrop_T(mob/target, mob/user)
-	if(!CanMouseDrop(target, user))
-		return
+/obj/machinery/atmospherics/unary/cryo_cell/proc/user_can_move_target_inside(mob/target, mob/user)
```
Ditto on user and warnings

SierraKomodo commented on code/game/machinery/cryopod.dm:
```diff
@@ -437,45 +437,57 @@
+	target.remove_grabs()
+	return TRUE
+
+/obj/machinery/cryopod/proc/user_can_move_target_inside(mob/target, mob/user)
```
And ditto here too

SierraKomodo commented on code/game/machinery/cryopod.dm:
```diff
@@ -437,45 +437,57 @@
-		return
-	if (QDELETED(target))
-		return
+		return FALSE
```
Inconsistent return values in this proc

emmanuelbassil commented:
Fixed what you requested.
Also added same stuff for surgery table; made patients face upwards, and updated remove_grab proc to remove pulls too.

SierraKomodo commented on code/game/machinery/OpTable.dm:
```diff
@@ -156,11 +145,20 @@
+	for (var/obj/item/grab/grab in target.grabbed_by)
+		if (grab.assailant == target || grab.assailant == user)
+			continue
+		USE_FEEDBACK_FAILURE("\The [target] is being grabbed by [grab.assailant] and can't be placed on \the [src].")
```
```suggestion
		USE_FEEDBACK_FAILURE("\The [target] is being grabbed by \the [grab.assailant] and can't be placed on \the [src].")
```

SierraKomodo replied:
```suggestion
		USE_FEEDBACK_FAILURE("\The [target] is being grabbed by \the [grab.assailant] and can't be placed on \the [src].")
```

SierraKomodo commented on code/game/machinery/Sleeper.dm:
```diff
@@ -183,18 +183,47 @@
 		return TRUE
 	return ..()
 
+/obj/machinery/sleeper/proc/user_can_move_target_inside(mob/target, mob/user)
```
This is a good candidate to add a `user.use_sanity_check(src, target)` check as well.

SierraKomodo commented on code/game/machinery/cryopod.dm:
```diff
@@ -419,64 +419,74 @@
 /obj/machinery/cryopod/proc/attempt_enter(mob/target, mob/user)
```
You've added return values to this proc, but there's no calls to it that use a return value.

SierraKomodo commented on code/game/machinery/cryopod.dm:
```diff
@@ -419,64 +419,74 @@
+	return TRUE
 
-//Like grap-put, but for mouse-drop.
 /obj/machinery/cryopod/MouseDrop_T(mob/target, mob/user)
```
`MouseDrop_T` doesn't use return values either.

SierraKomodo commented on code/game/objects/buckling.dm:
```diff
@@ -86,8 +86,9 @@
-			grabbed_by_mobs += grab.assailant
+		if (grab.assailant == user || grab.assailant == target)
+			continue
+		grabbed_by_mobs += grab.assailant
```
```suggestion
		grabbed_by_mobs += "\the [grab.assailant]"
```

emmanuelbassil committed:
use_grab on medical machines.

Also made it check for other people grabbing, made patients face upwards.
Test


Ability to use grab on sleeper.


QOL sleeper


Auto stash before rebase of "refs/heads/MedQI"
Auto stash before rebase of "refs/heads/MedQI"
Auto stash before rebase of "refs/heads/MedQI"
Auto stash before rebase of "refs/heads/MedQI"
Auto stash before rebase of "refs/heads/MedQI"
Auto stash before rebase of "refs/heads/MedQI"
Spacing


Fixes


Fixes


Fixes


Fixes


Fixes


Fixes


Fixes

emmanuelbassil commented:
Done!
